[Merged by Bors] - feat: custom elaborators for differential geometry

grunweg opened the pull request:
Add two sets of custom elaborators for differential geometry, to
- infer the model with corners in `MDifferentiable{WithinAt,At,On}` or `ContMDiff{WithinAt,At,On}` expressions from the local context, and provide shorter notation in that setting
- do the same for mfderiv: `mfderiv% f` is shorthand for `mfderiv I J f`; `mfderiv[s] f x` is short for `mfderivWithin I J f s`
- convert a fibre bundle section from a dependent section to a function into the total space, making working with differentiability of sections of fibre bundles less cumbersome

This was [discussed on zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Differential.20geometry.20elaborators.20experiment/with/528035295), to overall broad consensus.

This code has substantial tests (which have caught bugs in the past). It has also been tested in the branch in #26221 (uncovering minor issues, but not substantial ones --- the worst of which have been fixed).
Many thanks go to `fpvandoorn` (for lots of in-person comments on improving the meta code), and to `thorimur` for rewriting the entire implementation from ground up to be much more robust and maintainable.

A number of tweaks and extensions can be made in future PRs: this set of elaborators is useful already.
Once it is deemed sufficiently robust, differential geometry files can be converted to use it.

From the path towards geodesics and the Levi-Civita connection.

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: custom elaborators for differential geometry

Add two sets of custom elaborators for differential geometry, to
- infer the model with corners in `MDifferentiable{WithinAt,At,On}` or `ContMDiff{WithinAt,At,On}`
  expressions from the local context, and provide shorter notation in that setting
- convert a fibre bundle section from a dependent section to a function into the total space,
  making working with differentiability of sections of fibre bundles less cumbersome

This was [discussed on zulip](...), to overall broad consensus.

This code has substantial tests and has been tried in the branch in #26221 (uncovering minor issues,
but not substantial ones). The implementation itself is of prototype quality and will be cleaned up
in the future. Once it is more robust, more differential geometry files can be converted to use it.

-----------

I agree that merging not fully polished code is not ideal, but propose doing so for two reasons:
- the work in #26221 used these elaborators extensively (to test them).
  Reverting those changes, merging the code, only to re-apply them is not a good use of time.
  Merging this PR now enables merging the work in #26221 soon.
- the feature is clearly useful and not going away: in the files that use it, it works well already.
  Holding off on further adopting, but merging it now seems like a good compromise.

sgouezel commented:
Thanks for this! I agree that this is extremely useful, and should be merged quickly if that's possible. However, I think that the scoping to the `Manifold` namespace is really important, and should be included in the current PR.

grunweg commented:
Sure, I'm happy to do this - can you point me at how?

grunweg commented:
Trying the obvious things `elab[Manifold]:max "CMDiff" nt:term:arg t:term:arg : term => do ...` and `elab:max[Manifold] "CMDiff" nt:term:arg t:term:arg : term => do` did not work.

grunweg committed:
Remove unused deprecated elaborators; noshake

grunweg committed:
Scope delaborators

grunweg committed:
More minimal imports

grunweg committed:
Use double backticks

grunweg committed:
Manual noshake

grunweg committed:
chore(Elaborators): fix a few typos in doc-strings

grunweg committed:
feat: custom elaborator for mderiv(Within)

grunweg committed:
And add some tests, including some for (already fine) error messages

grunweg committed:
Fix the build

grunweg committed:
A few fixes from review

grunweg committed:
fix: precendence in T%.

grunweg committed:
chore: note one TODO
It's hard to make a test for this, though!

grunweg committed:
chore: review comments
Check reducible defeq instead. TODO, add tests for this!

grunweg committed:
wip: test for beta-reduction in T %

grunweg committed:
wip: beta-reduce in T% elaborator

grunweg committed:
Revert "wip: beta-reduce in T% elaborator"

This reverts commit c10258295e9626597b380f36b3dc7efaf0794b4e.

grunweg committed:
Two more typos

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

grunweg commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```diff
@@ -0,0 +1,390 @@
+
+In each of these cases, the models with corners are inferred from the domain and codomain of `f`.
+The search for models with corners uses the local context and is (almost) only syntactic, hence
+hopefully fast enough to always run.
```
At the moment, these elaborators don't really support products yet: I'll add that a future TODO!
(In these cases, indeed are are several choices - in particular for products of vector spaces. Yet, my impression was that picking the `ModelWithCorners.prod` one is usually better, and how most lemmas in mathlib are phrased. What do you think? (In any case, the elaborator should warn on a product of vector spaces, to alert the user of this.)

grunweg replied:
Turns out I already had a comment mentioning it; I just extended it.

grunweg reviewed (commented):
Thanks for the comments; let me look into them now!

grunweg committed:
Document trace classes

grunweg commented on Mathlib/Geometry/Manifold/Traces.lean:
```diff
@@ -0,0 +1,17 @@
+/-!
+# Trace classes for differential geometry elaborators
+
+TODO: add a more complete doc-string
```
Done; would you like me to add anything else?

grunweg committed:
Document product case better

grunweg committed:
Clarify cryptic TODO

grunweg committed:
chore: structure the tests better

grunweg committed:
chore: add many tests for the CMDiffAt elaborators

grunweg committed:
chore: add tests about coercion behaviour

grunweg commented on Mathlib/Geometry/Manifold/Elaborators.lean:
Sure. I just added a few tests, and will decide on this later this morning.

grunweg committed:
Support coercion n to WithTop ENat everywhere.

grunweg committed:
fix: guard against loose bvars, i.e. non-dependent functions

This fixes the cryptic 'unknown identifier #0' errors I saw previous,
and provides a *much* nicer error message.

grunweg committed:
Add tests for mfderiv and dependent fns also

The dependent function warning was added in the previous commit.

grunweg committed:
wip/refactor: use forallBoundedTelescope instead of manual matching

As discussed in-person with Floris: this should be more robust, but currently
fails (for reasons not yet understood).

grunweg committed:
Note another TODO

grunweg committed:
And another

grunweg committed:
Update error messages for merge

grunweg committed:
One more check and test

grunweg committed:
chore: use Hint: instead of Note:
This matches usage in Lean core. We are not providing further information
about the error, but telling the user how they might fix it

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```diff
@@ -0,0 +1,460 @@
+hopefully fast enough to always run.
+
+This has no dedicated support for product manifolds (or product vector spaces) yet;
+adding this is left for future changes. (It would make need to make a choice between e.g. the
```
```suggestion
adding this is left for future changes. (It would need to make a choice between e.g. the
```

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```diff
@@ -0,0 +1,460 @@
+This has no dedicated support for product manifolds (or product vector spaces) yet;
+adding this is left for future changes. (It would make need to make a choice between e.g. the
+trivial model with corners on a product `E × F` and the product of the trivial models on `E` and
+`F`).
```
Indicate that in these settings the elaborators should be avoided?

grunweg replied:
Good idea, done!

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```diff
@@ -0,0 +1,460 @@
+trivial model with corners on a product `E × F` and the product of the trivial models on `E` and
+`F`).
+
+Secondly, this space adds an elaborator to ease working with sections in a fibre bundle,
```
```suggestion
Secondly, this file adds an elaborator to ease working with sections in a fibre bundle,
```

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
The question still stands.

grunweg replied:
```suggestion
- add tests for all of the above
```

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
It's strange to have commented out code in the PR.

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
commented out?

sgouezel commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```suggestion
/-- `mfderiv[u] f x` elaborates to `mfderivWithin I J f u x`,
```

sgouezel commented on MathlibTest/DifferentialGeometry/Elaborators.lean:
```diff
@@ -0,0 +1,911 @@
+import Mathlib.Geometry.Manifold.Elaborators
+
```
```suggestion
```

sgouezel reviewed (commented):
I'm very happy with the extensive testing. I can't really review the meta code, but I'd also want ot get this in nevertheless. Can you bring this again on Zulip, asking for a quick look of a meta expert? If you do that but nobody shows up, I'll merge this in one week (don't hesitate to ping me!)

grunweg committed:
Apply suggestions from code review

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

grunweg committed:
Update MathlibTest/DifferentialGeometry/Elaborators.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

grunweg committed:
Stronger warning

grunweg commented:
Thank you! I asked on zulip (and pinged Floris specifically, as we talked about this a lot already).

grunweg committed:
Use Hint more

grunweg committed:
Document and polish find_models better

grunweg committed:
refactor: reduce duplication in the various small elaborators

grunweg committed:
chore: validate the types of s and src match

This caught one typo, so this was useful!

grunweg committed:
fix(Elaborators): pretty-print the new elaborated notation correctly

thorimur committed:
style: remove `>>=`

thorimur committed:
chore: `whnfR` after `instantiateMVars`

thorimur committed:
chore: remove `getUniverse` in favor of `mkAppOptM`

thorimur committed:
chore: `PrettyPrinter.delab` -> `Term.exprToSyntax`

thorimur committed:
chore(Elaborators): change trace class names

thorimur committed:
feat: move trace classes into main file; make collective trace class; make descriptions into docstrings

thorimur committed:
chore: rename `find_model` -> `findModel`, `_find_models` -> `findModels`

thorimur committed:
chore: comment out (fixed) `MDiffAt2` and remove test

thorimur committed:
chore: use Qq to construct `WithTop ℕ∞`

thorimur committed:
refactor: factor out loops into (private) definitions for readability; search local instances for classes instead of whole context

thorimur committed:
chore: remove unnecessary `m!`s

thorimur committed:
style: use `withLocalDeclD`

thorimur committed:
chore: append `Elab` namespace

thorimur committed:
style: fix `findModel` body indentation

thorimur committed:
chore: grammar

thorimur committed:
style: code fencing instead of quotes in messages

thorimur committed:
chore: typo

thorimur committed:
fix: don't use `BEq` to compare expressions

thorimur committed:
fix: do not use `arg` precedence in brackets

thorimur committed:
fix: reset `TermElabM` state in `findModel` appropriately when trying different strategies; refactor `findModel` to  make this possible.

* Note that the control flow changes slightly as a result of this commit. `TotalSpace`s will not cause top-level errors when both `TotalSpace` strategies fail; rather, the remaining strategies will be tried just in case. The given error message is demoted to a trace message. Better error messages for `findModel` can be provided in subsequent PRs.

thorimur committed:
chore: use `match_expr` in `findModel`

thorimur committed:
chore: use `match_expr` in `T%`

thorimur committed:
style: use dot notation for `hasLooseBVars`

thorimur committed:
chore: use `findSomeLocalInstanceOf?` in `T%`, and explain why we don't use `match_expr`

thorimur committed:
chore: move `findSomeLocalInstanceOf?` above `T%`

thorimur committed:
chore: perform cheapest check first

thorimur committed:
chore: use `pureIsDefEq` or explain why not

thorimur committed:
chore: use commented-out code instead of syntax re-elab

thorimur committed:
chore: refactor `findModels` to clean up elabs

thorimur committed:
chore: TODO comments

thorimur committed:
chore: remove `mkApp12`

thorimur committed:
fix: make elaborators `scoped`, as promised

grunweg committed:
chore: move mkApp12 to a Lean

grunweg committed:
chore: move getUniverse to Lean/Meta/ElabTerm

grunweg committed:
Fix

grunweg committed:
Fix test; tweak error message
XXX do I like the new defeq check, or are other checks better?

grunweg committed:
Merge branch 'master' into diffgeo-custom-elaborators

thorimur committed:
(Git wrangling) Revert "Fix test; tweak error message"

This reverts commit b3bf6090fffd577afb9a4f39c02d5b32d64f9f0b.

thorimur committed:
(Git wrangling) Revert "Fix"

This reverts commit 3abe5ed16c5eff02947f01a4db81a283c74c90d4.

thorimur committed:
(Git wrangling) Revert "chore: move getUniverse to Lean/Meta/ElabTerm"

This reverts commit 74cfff3f867da83b158eb34628a08ef72bb38354.

thorimur committed:
(Git wrangling) Revert "chore: move mkApp12 to a Lean"

This reverts commit 52e0ee02133affaaf250e9535e5a91978029429a.

thorimur committed:
docs: clean up module docstring

thorimur committed:
docs: misc. style, typos, wording

thorimur committed:
style(`T%`): `e.app`

thorimur committed:
style: line reflow

thorimur committed:
chore: comment for `findSomeLocalInstanceOf?`

thorimur committed:
fix: scope `T%` to `Manifold`

thorimur committed:
chore: reduce `etype` to get `.forallE`

thorimur committed:
style: indent second line of multiline errors

thorimur committed:
chore: improve error message for trivial bundle

thorimur committed:
test: tracing

thorimur committed:
chore: affirm that this check is fine!

thorimur reviewed (commented):
First, thanks for writing this PR! :) It was very nice to review.

This is a large review, with a large number of suggested changes, which I've put into [this PR](https://github.com/grunweg/mathlib4/pull/1) to avoid stressing the Git suggested changes interface.

Each commit is intended to be a readable suggested change; the number of comments here comes from thoroughness in explaining those changes, rather than representing a huge number of requests! :) In fact, the only requests per se (besides considering the suggested changes) are detailed at the bottom of this comment.

Given that the design is already finalized, I only reviewed:
- the implementation and metaprogramming choices
- docstrings
- test coverage
- style

The implementation changes I'm suggesting are based on metaprogramming good practice and maintainability; see especially comments around `findModel` and `findModels`. (`T%` sees some changes too.) I avoided changes to the algorithm itself, and only suggested modifications to the implementation to ensure it behaved well, fulfilled metaprogramming responsibilities, and was readable.

I have two points I'd like to raise here:

* The test coverage looks great! I only noticed one hole: could you please add a test which finds a model from a `TotalSpace` of a tangent bundle? The `Bundle.TotalSpace _ (TangentSpace ..)` case seemed to be the only one that was not explored by at least one test. (Adding this test after merging suggested changes will probably be easiest for git!)

* I would also like to consider this file's name and location. I notice that no other file in Mathlib is named `Elaborators`. Do you know if this is just because differential geometry is (i.e. will be) unique in having several similar elaborators which are complex enough to warrant their own file?

grunweg commented:
Thanks a lot for this thorough and detailed review, this is extremely helpful! I responded on https://github.com/grunweg/mathlib4/pull/1 with two small questions; apart from those, the code looks great to me.

grunweg committed:
Merge pull request #1 from thorimur/diffgeo-custom-elaborators-review

review: suggested changes for #27021

grunweg commented:
Copy-pasting my comment from grunweg#1:
- can you say something about [0f85d742a78403c9676447789d18a0a4171dc6af](https://github.com/grunweg/mathlib4/pull/1/commits/0f85d742a78403c9676447789d18a0a4171dc6af)? If it works, I'm happy to blindly follow you, but I don't understand why this is superfluous.
- do you have a reference where I can read about `match_expr`? The language reference doesn't mention it, and the API docs seemingly also don't.
- I'm a bit surprised by [95f24aed78a7e17101c6d0d4f67a0f6e146ca96a](https://github.com/grunweg/mathlib4/pull/1/commits/95f24aed78a7e17101c6d0d4f67a0f6e146ca96a): is that really Lean core style?

grunweg commented on Mathlib/Geometry/Manifold/Elaborators.lean:
```suggestion
- extend the elaborators to support `OpenPartialHomeomorph`s and `PartialEquiv`s
```

grunweg committed:
chore: move file to Notation.lean

grunweg commented:
I just tried to add the missing test, and things failed in an interesting way:

```
-- Inferring a model with corners on the trivial bundle over the tangent space of a vector bundle.
set_option trace.Elab.DiffGeo true

/--
error: Could not find models with corners for TotalSpace F (TangentSpace I)
---
trace: [Elab.DiffGeo.MDiff] Finding a model for: TotalSpace F (TangentSpace I)
[Elab.DiffGeo.MDiff] ❌️ TotalSpace
  [Elab.DiffGeo.MDiff] ❌️ TangentSpace
    [Elab.DiffGeo.MDiff] Failed with error:
        TangentSpace I is not a `TangentSpace`
  [Elab.DiffGeo.MDiff] ❌️ From base info
    [Elab.DiffGeo.MDiff] Failed with error:
        No `baseInfo` provided
  [Elab.DiffGeo.MDiff] Failed with error:
      Having a TotalSpace as source is not yet supported
[Elab.DiffGeo.MDiff] ❌️ NormedSpace
  [Elab.DiffGeo.MDiff] Failed with error:
      Couldn't find a `NormedSpace` structure on TotalSpace F (TangentSpace I) among local instances.
[Elab.DiffGeo.MDiff] ❌️ ChartedSpace
  [Elab.DiffGeo.MDiff] Failed with error:
      Couldn't find a `ChartedSpace` structure on TotalSpace F (TangentSpace I) among local instances.
[Elab.DiffGeo.MDiff] ❌️ NormedField
  [Elab.DiffGeo.MDiff] Failed with error:
      failed to synthesize
        NontriviallyNormedField (TotalSpace F (TangentSpace I))
      ⏎
      Hint: Additional diagnostic information may be available using the `set_option diagnostics true` command.
-/
#guard_msgs in
variable {h : Bundle.TotalSpace F (TangentSpace I : M → Type _) → F} in
#check MDiff h
```

@thorimur What do you think?

grunweg committed:
Review comments

grunweg committed:
chore: add missing test; fails for reasons I don't understand

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
Apologies, I didn't add docstrings to these auxiliary definitions. :)
```suggestion
where
  /- Note that errors thrown in the following are caught by `tryStrategy` and converted to trace
  messages. -/
  /-- Attempt to find a model from a `TotalSpace` first by seeing if it is the total space of a 
  tangent bundle, then by attempting to use any provided `baseInfo`. -/
  fromTotalSpace : TermElabM Expr := do
    match_expr e with
    | Bundle.TotalSpace _ F V => do
      if let some m ← tryStrategy m!"TangentSpace" (fromTotalSpace.tangentSpace V) then return m
      if let some m ← tryStrategy m!"From base info" (fromTotalSpace.fromBaseInfo F) then return m
      throwError "Having a TotalSpace as source is not yet supported"
    | _ => throwError "{e} is not a `Bundle.TotalSpace`."
  /-- Attempt to find a model from the total space of a tangent bundle. -/
  fromTotalSpace.tangentSpace (V : Expr) : TermElabM Expr := do
    match_expr V with
    | TangentSpace _k _ _E _ _ _H _ I M _ _ => do
      trace[Elab.DiffGeo.MDiff] "This is the total space of the tangent bundle of {M}"
      let srcIT : Term ← Term.exprToSyntax I
      let resTerm : Term ← ``(ModelWithCorners.prod $srcIT (ModelWithCorners.tangent $srcIT))
      Term.elabTerm resTerm none
    | _ => throwError "{V} is not a `TangentSpace`"
  /-- Attempt to use the provided `baseInfo` to find a model. -/
  fromTotalSpace.fromBaseInfo (F : Expr) : TermElabM Expr := do
    if let some (src, srcI) := baseInfo then
      trace[Elab.DiffGeo.MDiff] "Using base info {src}, {srcI}"
      let some K ← findSomeLocalInstanceOf? ``NormedSpace fun _ type ↦ do
          match_expr type with
          | NormedSpace K E _ _ =>
            if ← withReducible (pureIsDefEq E F) then return some K else return none
          | _ => return none
        | throwError "Couldn't find a `NormedSpace` structure on {F} among local instances."
      trace[Elab.DiffGeo.MDiff] "{F} is a normed field over {K}"
      let kT : Term ← Term.exprToSyntax K
      let srcIT : Term ← Term.exprToSyntax srcI
      let FT : Term ← Term.exprToSyntax F
      let iTerm : Term ← ``(ModelWithCorners.prod $srcIT 𝓘($kT, $FT))
      Term.elabTerm iTerm none
    else
      throwError "No `baseInfo` provided"
  /-- Attempt to find the trivial model on a normed space. -/
  fromNormedSpace : TermElabM Expr := do
    let some (inst, K) ← findSomeLocalInstanceOf? ``NormedSpace fun inst type ↦ do
        match_expr type with
        | NormedSpace K E _ _ =>
          if ← withReducible (pureIsDefEq E e) then return some (inst, K) else return none
        | _ => return none
      | throwError "Couldn't find a `NormedSpace` structure on {e} among local instances."
    trace[Elab.DiffGeo.MDiff] "Field is: {K}"
    mkAppOptM ``modelWithCornersSelf #[K, none, e, none, inst]
  /-- Attempt to find a model with corners on a manifold. -/
  fromChartedSpace : TermElabM Expr := do
    let some H ← findSomeLocalInstanceOf? ``ChartedSpace fun _ type ↦ do
        match_expr type with
        | ChartedSpace H _ M _ =>
          if ← withReducible (pureIsDefEq M e) then return some H else return none
        | _ => return none
      | throwError "Couldn't find a `ChartedSpace` structure on {e} among local instances."
    trace[Elab.DiffGeo.MDiff] "H is: {H}"
    let some m ← findSomeLocalHyp? fun fvar type ↦ do
        match_expr type with
        | ModelWithCorners _ _ _ _ _ H' _ => do
          if ← withReducible (pureIsDefEq H' H) then return some fvar else return none
        | _ => return none
      | throwError "Couldn't find a `ModelWithCorners` with model space {H} in the local context."
    return m
  /-- Attempt to find a model with corners from a normed field. We attempt to find a global instance here. -/
  fromNormedField : TermElabM Expr := do
```

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
I wasn't too sure about it either; happy to see it change in the future. :) (Aside: I looked for precedent, but trace class names don't seem to be standardized yet! Maybe we'd like to change this.)

thorimur commented:
> do you have a reference where I can read about match_expr? The language reference doesn't mention it, and the API docs seemingly also don't.

`match_expr` is indeed sorely in need of documentation, unfortunately. :( (It doesn't even have a docstring!)

You can see a bit of what it does, though, by printing a definition which uses it:
```
def foo (e : Expr) : MetaM Unit := do
  match_expr e with
  | Not _ => return
  | _ => return

#print foo
/-
def Manifold.Elab.foo : Expr → MetaM Unit :=
fun e ↦ do
  let discr ← instantiateMVarsIfMVarApp e
  have __do_jp : Unit → MetaM Unit := fun x ↦ pure PUnit.unit
  have __do_jp_1 : Unit → MetaM Unit := fun x ↦ pure PUnit.unit
  have __discr : Expr := discr.cleanupAnnotations
  if h : __discr.isApp = true then
      have __discr := __discr.appFnCleanup h;
      if __discr.isConstOf `Not = true then __do_jp () else __do_jp_1 ()
    else __do_jp_1 ()
-/
```

You can see that it does all sorts of nice preprocessing for `e` at various stages (notably `instantiateMVarsIfMVarApp`, `cleanupAnnotations`, and `appFnCleanup`. (It's still a little weaker than `whnfR`, but this is a good thing, since we don't always want to see through reducibility.)

I do wish there were some resource that detailed it further, though. (If I have time I might try to contribute documentation myself...)

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
I reflected on my suggestion to use `pureIsDefEq` here, and changed my mind: I think `isDefEq` is appropriate. We might actually even like to be more aggressive; I've left that as a TODO.

```suggestion
      /- Note: we use `isDefEq` here since persistent metavariable assignments in `src` and 
      `estype` are acceptable.
      TODO: consider attempting to coerce `es` to a `Set`. -/
      if !(← isDefEq estype <|← mkAppM ``Set #[src]) then
```

grunweg committed:
Update Mathlib/Geometry/Manifold/Notation.lean

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Add docs for auxiliary doc-strings

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

grunweg commented:
> > do you have a reference where I can read about match_expr? The language reference doesn't mention it, and the API docs seemingly also don't.
> 
> `match_expr` is indeed sorely in need of documentation, unfortunately. :( (It doesn't even have a docstring!)
> 
> You can see a bit of what it does, though, by printing a definition which uses it:
> 
> ```
> def foo (e : Expr) : MetaM Unit := do
>   match_expr e with
>   | Not _ => return
>   | _ => return
> 
> #print foo
> /-
> def Manifold.Elab.foo : Expr → MetaM Unit :=
> fun e ↦ do
>   let discr ← instantiateMVarsIfMVarApp e
>   have __do_jp : Unit → MetaM Unit := fun x ↦ pure PUnit.unit
>   have __do_jp_1 : Unit → MetaM Unit := fun x ↦ pure PUnit.unit
>   have __discr : Expr := discr.cleanupAnnotations
>   if h : __discr.isApp = true then
>       have __discr := __discr.appFnCleanup h;
>       if __discr.isConstOf `Not = true then __do_jp () else __do_jp_1 ()
>     else __do_jp_1 ()
> -/
> ```
> 
> You can see that it does all sorts of nice preprocessing for `e` at various stages (notably `instantiateMVarsIfMVarApp`, `cleanupAnnotations`, and `appFnCleanup`. (It's still a little weaker than `whnfR`, but this is a good thing, since we don't always want to see through reducibility.)
> 
> I do wish I knew of some resource that detailed it further, though. (If I have time I might try to contribute documentation myself...) (I first saw it when Leo presented it at some virtual presentation, though I'm not sure which one. Possibly LeanTogether 2025?)

Thanks for elaborating! This makes a lot of sense now. (And having it documented would indeed be great.)

grunweg committed:
Update expected output

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
It seems that we were relying on the `TangentSpace` match to fail in order to use the (correct) `baseInfo`! It seems like `baseInfo` should actually override the `TangentSpace` match, as far as I can tell.

(This doesn't fix the `TopologicalSpace` synthesis failure.)

```suggestion
  /-- Attempt to find a model from a `TotalSpace` first by attempting to use any provided
  `baseInfo`, then by seeing if it is the total space of a tangent bundle. -/
  fromTotalSpace : TermElabM Expr := do
    match_expr e with
    | Bundle.TotalSpace _ F V => do
      if let some m ← tryStrategy m!"From base info" (fromTotalSpace.fromBaseInfo F) then return m
      if let some m ← tryStrategy m!"TangentSpace" (fromTotalSpace.tangentSpace V) then return m
      throwError "Having a TotalSpace as source is not yet supported"
    | _ => throwError "{e} is not a `Bundle.TotalSpace`."
  /-- Attempt to use the provided `baseInfo` to find a model. -/
  fromTotalSpace.fromBaseInfo (F : Expr) : TermElabM Expr := do
    if let some (src, srcI) := baseInfo then
      trace[Elab.DiffGeo.MDiff] "Using base info {src}, {srcI}"
      let some K ← findSomeLocalInstanceOf? ``NormedSpace fun _ type ↦ do
          match_expr type with
          | NormedSpace K E _ _ =>
            if ← withReducible (pureIsDefEq E F) then return some K else return none
          | _ => return none
        | throwError "Couldn't find a `NormedSpace` structure on {F} among local instances."
      trace[Elab.DiffGeo.MDiff] "{F} is a normed field over {K}"
      let kT : Term ← Term.exprToSyntax K
      let srcIT : Term ← Term.exprToSyntax srcI
      let FT : Term ← Term.exprToSyntax F
      let iTerm : Term ← ``(ModelWithCorners.prod $srcIT 𝓘($kT, $FT))
      Term.elabTerm iTerm none
    else
      throwError "No `baseInfo` provided"
  /-- Attempt to find a model from the total space of a tangent bundle. -/
  fromTotalSpace.tangentSpace (V : Expr) : TermElabM Expr := do
    match_expr V with
    | TangentSpace _k _ _E _ _ _H _ I M _ _ => do
      trace[Elab.DiffGeo.MDiff] "This is the total space of the tangent bundle of {M}"
      let srcIT : Term ← Term.exprToSyntax I
      let resTerm : Term ← ``(ModelWithCorners.prod $srcIT (ModelWithCorners.tangent $srcIT))
      Term.elabTerm resTerm none
    | _ => throwError "{V} is not a `TangentSpace`"
```

grunweg committed:
Update Mathlib/Geometry/Manifold/Notation.lean

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Comment the recent test

grunweg committed:
chore: move test

grunweg committed:
chore: add regression test for precendence in the "on/within" set

grunweg committed:
Clarify failing test, and document the right spelling with TangentBundle

grunweg commented:
> Hmm, relatedly, I notice that `TangentBundle` is an `abbrev` for `Bundle.TotalSpace _ (TangentSpace _)`. Should `findModel` see `TangentBundle` as a `Bundle.TotalSpace`, or handle it as a separate case (or not handle it)? If the former, then we should use `match_expr ← whnfR e with` in `fromTotalSpace` to unwrap `TangentBundle`. But if it has separate instances, it might make sense to handle it as a separate case if at all (presumably in a subsequent PR!).
> 
> This is a question of the algorithm's spec, so I'll defer to those with greater library knowledge in this area. :)

I think that
- the elaborators should definitely support `TangentBundle` (either now, or in a follow-up)
- `TangentBundle` has separate instances, but not about manifolds... so I wondered if just changing the `match_expr` suffices. Trying it out, the answer seems to be no. So, it might be best to add a new case to `findModel`, and perhaps do so in a follow-up.

grunweg commented:
In other words: from my side, this PR is ready to get merged. There are a number of folliow-up changes I'd like to make, but these can be subsequent PRs.

grunweg committed:
Noshake; fix bad merge/import

grunweg commented:
CI passes on this PR. I've gone over the diff again, and also resolved all stale review comments. I'll keep Thomas' more substantial meta comments open (for future me to re-read them); they are all addressed. All in all, this seems ready for merging (but I'm of course biased).

@thorimur Let me just repeat again that I'm really grateful for your extremely detailed review. I learned a lot (and will probably learn more as I return to your comments to re-process them). It was a pleasure, until next time/ one of the follow-up changes :-)
Would you like to add yourself to the authors of this file? I have no idea what mathlib's etiquette says, but you have certainly rewritten the meta code significantly.

thorimur commented:
I'm so glad my comments were useful to you! :) I'm always happy to talk about metaprogramming, so if you return to these comments later, feel free to ask me further questions about them (e.g. on Zulip)! :)

> Would you like to add yourself to the authors of this file? I have no idea what mathlib's etiquette says, but you have certainly rewritten the meta code significantly.

Hmm, I'm actually not sure if I should be on the author's list either!

I once heard someone say that the authors list is basically the answer to "who do I ping if I need to address something in this file", and from that perspective, I'd be happy to take on the responsibility of being among those pinged.

On the other hand, I'm not sure if it's a great idea to set a precedent of receiving authorship credit as a reviewer.

I'm perfectly okay with either outcome, but I'd rather defer to an opinion from a maintainer on what should happen here than ask to be there myself.

(If this turns out to be complicated, though, perhaps let's leave me off and not hold up the PR any further—I'd like to stay involved with these elaborators, so maybe I'll contribute to this file directly in the future and render the question moot. :) )

sgouezel commented:
bors d+
I think it definitely makes sense to add you as a second author to the file (not the least because then you will feel responsible for it, and then we can ping you again :-)

grunweg committed:
Update Mathlib/Geometry/Manifold/Notation.lean
